### ISSUE TITLE
Addon-registered alert frame subsystems can break actionbars

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#341]
Addons that register custom alert frame subsystems can trigger a chain of taint spread from the "What's New" splash frame, through the objective tracker, into edit mode, and onto the action bars. This leads to a case where actionbuttons become unusable via keybindings.

# Test Case

1. Run the following command: `/run AlertFrame:AddQueuedAlertFrameSubSystem("AlertFrameTemplate", nop, 1, 1)`
2. Open the game menu.
3. Click "What's New".
4. Close the "What's New" panel.
5. Enter edit mode.
6. Exit edit mode.
7. Cycle the current action bar page back and forth one place.
8. Press any keybound ability.

At this point you'll received a blocked action error blaming `*** ForceTaint_Strong ***` (the `/run` command). Step 7 reportedly may not be needed, but I've included it just in case.

# Explanation

The initial insecure call to [AlertFrame:AddQueuedAlertFrameSubsystem] creates and adds a new subsystem to the `self.alertFrameSubsystems` table, inserting it as a tainted entry.

Upon closing the "What's New" panel, [SplashFrame:OnHide] is called. This will call [AlertFrame:SetAlertsEnabled] which indirectly calls other functions that traverse the contents of the `self.alertFrameSubsystems` table, which will taint execution when reaching our tainted subsystem entry.

Once [AlertFrame:SetAlertsEnabled] returns, that taint then spreads into the ObjectiveTracker_Update call and later Edit Mode spreads it everywhere else due to the lack of system isolation described in #261.

# Solutions

**Re-order the sequence of calls in [SplashFrame:OnHide].**
At a surface level I believe there's no significance to the ordering of calls here, and updating the objective tracker before touching alert frame subsystems would solve this particular issue. It is however possible that there's an interaction that I'm missing which requires the existing ordering of calls be preserved.

```diff
diff --git a/Interface/FrameXML/SplashFrame.lua b/Interface/FrameXML/SplashFrame.lua
index 4a67a2ad..9e5b8584 100644
--- a/Interface/FrameXML/SplashFrame.lua
+++ b/Interface/FrameXML/SplashFrame.lua
@@ -21,8 +21,8 @@ end
 function SplashFrameMixin:OnHide()
 	self.screenInfo = nil;
 	C_TalkingHead.SetConversationsDeferred(false);
-	AlertFrame:SetAlertsEnabled(true, "splashFrame");
 	ObjectiveTracker_Update();
+	AlertFrame:SetAlertsEnabled(true, "splashFrame");
 end
 
 function SplashFrameMixin:OnEvent(event, ...)

```

**Deploy a [securecallfunction] barrier for toggling alert frames.**
Simple solution and probably safe enough; I doubt there's any gain from having taint from alert iteration propagate to the objective tracker.

```diff
diff --git a/Interface/FrameXML/SplashFrame.lua b/Interface/FrameXML/SplashFrame.lua
index 4a67a2ad..7760d87d 100644
--- a/Interface/FrameXML/SplashFrame.lua
+++ b/Interface/FrameXML/SplashFrame.lua
@@ -21,7 +21,7 @@ end
 function SplashFrameMixin:OnHide()
 	self.screenInfo = nil;
 	C_TalkingHead.SetConversationsDeferred(false);
-	AlertFrame:SetAlertsEnabled(true, "splashFrame");
+	securecallfunction(AlertFrame.SetAlertsEnabled, AlertFrame, true, "splashFrame");
 	ObjectiveTracker_Update();
 end
 ```

**Create an inbound attribute delegate for secure registration of alert frame subsystems.**
This is a much larger ask but it would possibly solve _other_ issues if we later find out that custom alert subsystems are spreading taint elsewhere.

[SplashFrame:OnHide]: https://github.com/Gethe/wow-ui-source/blob/28ab53f48c691a976daa33acb509e96d9b850503/Interface/FrameXML/SplashFrame.lua#L21
[AlertFrame:SetAlertsEnabled]: https://github.com/Gethe/wow-ui-source/blob/28ab53f48c691a976daa33acb509e96d9b850503/Interface/FrameXML/AlertFrames.lua#L291
[AlertFrame:AddQueuedAlertFrameSubsystem]: https://github.com/Gethe/wow-ui-source/blob/28ab53f48c691a976daa33acb509e96d9b850503/Interface/FrameXML/AlertFrames.lua#L330
[securecallfunction]: https://wowpedia.fandom.com/wiki/API_securecallfunction

**Comment by TomCats** [comment]
I would consider that the body of ObjectiveTracker_Update() itself always be executed via securecallfunction().  While it's a good idea to prevent the fields being updated from becoming tainted in the first place, such as through the suggestions from Meorawr, much damage would be mitigated by not having a secure execution path become insecure upon the return from this function.

You can do it better than this, but as an example of the concept:

function ObjectiveTracker_Update(reason, id, moduleWhoseCollapseChanged)
    securecall(function()
          ... stuff in ObjectiveTracker_Update...  Let it taint itself if it happens and we have 1 problem to solve instead of 1000 since it won't spread the taint to its caller ...
    )
end

**Comment by Meorawr** [comment]
That won't fix anything if execution is tainted prior to entry into that function, which is what's happening here.

**Comment by TomCats** [comment]
Whether or not we want to prevent a tainted process from calling ObjectiveTracker_Update is only one half of the concern.  Perhaps there are features within ObjectiveTracker that would be affected by this taint and perhaps there are not.  If there are, certainly preventing the taint from reaching the ObjectiveTracker_Update call is important.  You reported this to trigger issues within the Edit mode, therefore, the underlying problem is that any tainted call to ObjectiveTracker_Update will spread taint to any other process which subsequently calls it.

While your solution to prevent the one example of taint spread is genuinely valid and should be implemented ASAP, it doesn't alone solve the bigger problem; nor does my solution alone solve the problem of the Objective Tracker becoming tainted.  Blizzard should do both together...  else SplashFrame isn't going to be the last trigger and probably isn't the only trigger today.

**Comment by Meorawr** [comment]
That's also an entirely separate issue. I don't think that preventing tainting leaking out of the objective tracker into other systems is the correct play here; rather Edit Mode itself should just be processing systems in isolation via `secureexecuterange` (tangentially discussed in #261). That way the other issues of FCF taint (#309) and UI scale modification (#303) bricking edit mode are more centrally solved.

As for this issue specifically - that this report points out the Edit Mode interaction is more just that it's a _convenient_ way of visibly reproducing the issue. The sole aim of this report is simply for alert frames to not taint the objective tracker.

**Comment by TomCats** [comment]
I'll be eager to see a resolution either way.  TomCat's Tours was affected by this specific issue and I am exploring a work-around that is better than disabling its use of the AlertFrame

**Comment by Meorawr** [comment]
Should be resolved in 10.0.2.47067.
